Sharing plows

kevingranade opened the pull request:
A few tweaks on top of #13376 
Bicycles could operate seed drills and reapers, and cars could pull multiple plows at high speeds, which seems to be fixed now.
The model for drag is simple, the accessory states how much drag it imposes as negative power, which is accumulated in vehicle::refresh() and exposed by vehicle::drag().  This drag number is added to the existing drag amount in vehproceed _after_ any engine thrusting for the turn is applied, so if the drag exceeds the power output of the vehicle's engines, it simply doesn't move, but the engine does sit there revving, which is about right.
I tested with a regular car with two plows attached, a bicycle with one plow attached, and the plow tractor, all of them operate as expected, except the tractor has a top speed just under 500MPh, but that's another issue :/

I think it's good to go now, let me know if I missed anything.

Fixed issues:
- [x] Error messages when total_power() goes negative, because the vehicle code thinks that means there is no engine.
- [x] No deceleration when drag is greater than engine power, you can get a bike to 20MPH and then engage the plow, which will plow a nice furrow for you.
- [x] Have to release controls and re-grab them for plow deactivation to be noticed.

Remaining issues:
- [ ] No notification when your vehicle is unable to move due to the engine not being strong enough to overcome the attached accessories.
- [ ] No strain on engine when barely able to plow, though it does burn a lot of fuel.

88c4442 fixes a bug where power consumption wasn't updated after toggling things on and off with the controls.

jaked122 commented:
Neat. So, I see that's what you meant. You wanted the refresh method to take care of adding additional drag when the tools were turned on. I hadn't thought of calling it at the end of use_controls. 

That does make more sense than adding a static amount of engine resistance to the tools.

Coolthulhu commented:
4 days and no updates. Are those two remaining issues supposed to be fixed here or in some future PRs?

kevingranade commented:
The delay is due to me being on vacation, I'd like to address the problems
before merging, but wanted to post my progress to gather any feedback and
as something for people to build on if they felt like picking it up and
addressing the issues.

kevingranade commented:
I don't think the remaining issues are compelling problems, this works now.

BevapDin commented on doc/JSON_FLAGS.md:
````diff
@@ -430,6 +430,12 @@ These branches are also the valid entries for the categories of `dreams` in `dre
+- ```REAPER``` Cuts down mature crops, depositing them on the square
+  - The ```bonus``` tag defines how productive the harvest can be.
+- ```PLOW``` Tills the soil underneath the part while active. Takes damage from unsuitable terrain at a level proportional to the speed of the vehicle.
+- ```EXTRA_DRAG``` tells the vehicle that the part exerts engine power reduction.
````
`EXTRA_DRAG` is documented and even used in JSON, but there is not a single reference to it in C++. The code their still checks for the PLOW/REAPER etc. flags to add to `vehicle::extra_drag`.

BevapDin commented on src/map.cpp:
```diff
@@ -620,6 +617,10 @@ void map::move_vehicle( vehicle &veh, const tripoint &dp, const tileray &facing
         }
+        // accept new position
+        // if submap changed, we need to process grid from the beginning.
+        // TODO: Allow vehicles to get displaced vertically
```
I don't get the comments here. What do they have to do with the new code? Displacing vertically is now possible and that line should probably go. Similar the other two (why even bring up the submaps?)

BevapDin commented on src/vehicle.cpp:
```diff
@@ -3815,8 +3857,88 @@ void vehicle::idle(bool on_map) {
     if( on_map ) {
         update_time();
-        if(scoop_on){
-            operate_scoop();
```
Moving `operate_scoop` into `on_move` makes the scoop only work when the vehicles moves by itself. Previously it would work when the vehicle was standing still or when it was pulled/pushed by the character. That behavior was quite useful and realistic.

BevapDin commented on src/vehicle.cpp:
```diff
@@ -3815,8 +3857,88 @@ void vehicle::idle(bool on_map) {
+        if( g->m.furn(reaper_pos) != f_plant_harvest ){
+            continue;
+        }
+        islot_seed &seed_data = *g->m.i_at(reaper_pos).front().type->seed;
```
The whole code about spewing the harvested plants looks familiar. Could this have been copied from `iexamine::aggie_plant`? It should be in a separate function that can be called from here and from `iexamine::aggie_plant`.

This here will also harvest marloss and fungal stuff with none of the side effects from `iexamine::aggie_plant`.

BevapDin commented on src/vehicle.cpp:
```diff
@@ -3815,8 +3857,88 @@ void vehicle::idle(bool on_map) {
+        const int plant_produced =  rng( 1, parts[reaper_id].info().bonus );
+        const int seed_produced = rng(1, 3);
+        item tmp;
+        const tripoint &reaper_pos = start_reaper + parts[reaper_id].precalc[0];
```
Beside the seemingly random usage of `const tripoint &` vs `const tripoint` (reference vs value, technically fine, but very confusing), what in Barry's name is that coordinate calculation supposed to be?

```C++
const tripoint start_reaper = veh_start + parts[reaper_id].precalc[0];
```
Ok, the absolute position of the reaper part in the map coordinate system.

```C++
const tripoint &reaper_pos = start_reaper + parts[reaper_id].precalc[0];
```
Huh? Why add `precalc[0]` again? This is btw. the only usage of `start_reaper`. The line is identical to
```C++
const tripoint &reaper_pos = veh_start + parts[reaper_id].precalc[0] + parts[reaper_id].precalc[0];
```

BevapDin commented on src/vehicle.cpp:
```diff
@@ -3815,8 +3857,88 @@ void vehicle::idle(bool on_map) {
+                } else if( !g->m.has_flag( "DIGGABLE", loc ) ) {
+                    //If it isn't diggable terrain, then it will most likely be damaged.
+                    damage( planter_id, rng(1, 10), DT_BASH, false );
+                    sounds::sound(global_pos3() + parts[planter_id].precalc[0], rng(10,20), _("Clink"));
```
The location has already been calculated and stored in `loc`, it hasn't changed and shouldn't be recalculated.

BevapDin commented on src/vehicle.cpp:
```diff
@@ -3815,8 +3857,88 @@ void vehicle::idle(bool on_map) {
+                }
+                i->bday = calendar::turn;
+                g->m.add_item(loc, *i);
+                i = v.erase(i);
```
Most seed items are counted by charges (hemp is one exception). This code will plant a full *stack* of seeds.

```C++
if( !i->count_by_charges() || i->charges == 1 ) {
    i->bday = calendar::turn; 
    g->m.add_item( loc, *i );
    v.erase( i );
} else { // count_by_charges && charges > 1
    item tmp = *i; // don't change the bday of the remaining stack.
    tmp.bday = calendar::turn;
    g->m.add_item( loc, tmp );
    i->charge--;
}
```
(Count-by-charges item with only one charge behave pretty much like not count-by-charges item.)

BevapDin commented on src/vehicle.cpp:
```diff
@@ -1057,10 +1071,19 @@ void vehicle::use_controls()
+        menu.addentry( toggle_plow, true, MENU_AUTOASSIGN, _("Toggle Plow"));
+    }
+    if( has_planter ){
+        menu.addentry( toggle_planter, true, 'P', _("Toggle Planter"));
```
There is a bit of inconsistency with the vehicle controls menu: some options say "toggle X", others say "enable X" or "disabled X", depending on the current state. The second version should probably be used for all options as it gives the player a way to check the current status. Currently there is now way for the player to check whether the plow or planer is on.

drbig commented:
I *do* want this in.

kevingranade commented:
Me too, I've been putting most of my non-merge related time into sorting
out my zombie speed overhaul (which morphed into keeping things the same
for the moment, but even that is proving difficult), so this is on the
backburner.  If someone else wants to pick this up and sort out the (very
valid IIRC) issues BevapDin raised, I wouldn't mind at all.
Otherwise, I *do* plan on revisiting this once the monster move speed thing
is sorted out.

chaosvolt commented on src/vehicle.cpp:
```diff
@@ -3815,8 +3857,88 @@ void vehicle::idle(bool on_map) {
+    }
+}
+
+void vehicle::operate_planter(){
```
Ahah. From the looks of it, this is where the madness begins. See https://github.com/CleverRaven/Cataclysm-DDA/pull/13652 and https://github.com/CleverRaven/Cataclysm-DDA/pull/13653 for more Fun.

kevingranade commented:
Neglected to rebuild after a final rebase of the branch after incorporating BevapDin's changes, sorry about that.

chaosvolt commented:
Seemed to have been swiftly fixed at least. Strangely, when I snagged a recent version to test it, it didn't break the actual game itself, just he pull requests.

I'd also initially assumed it was merely due to the older PRs not having been branched from master versions with these functions, before I noticed the duplicates. Merge conflicts are a inevitable thing to deal with, but this is the first time I've witnessed it occur on such a large scale. o3o
